pca, integration, kNN second draft (#10)
CellDiscoveryNetwork/workshops: beginners-guide-to-analyzing-scRNAseq/day-1/4-PCA_Harmony_kNN.qmd

```{r}
se <- se %>%
    NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(
        method = "vst",
        nfeatures = 3000,
        verbose = FALSE) %>%
    ScaleData(verbose = FALSE, features = VariableFeatures(.))
```

## Analysis

### PCA
When we run PCA on our Seurat object using the Seurat package's implementation, PCA is calculated on the scaled, highly variable genes. The resulting PCA latent space is placed in the *reductions* slot.

```{r}
se <- se %>% 
    RunPCA(
        npcs = 30,
        ndims.print = 1:5,
        nfeatures.print = 10
    )

# This is where latent spaces, aka dimensional reductions are stored. If we call it directly, we can see some information about the PCA.
se@reductions$pca
```

#### Accessing PCA results
The PC x Cell latent space is stored in the cell.embeddings slot.
The Gene x PC matrix of gene weights per PC is stored in the feature.loadings slot. 
The variance explained per PC is placed in the stdev slot
```{r}
DT::datatable(se@reductions$pca@cell.embeddings)

DT::datatable(se@reductions$pca@feature.loadings)

(se@reductions$pca@stdev) %>% head
```

#### Using PCA to understand variance in your data

> Just by visualizing these two matrices, we can start to understand the major axes of variance of our data. Seurat has built-in functions to graph the PCs and loadings

Reviewer: Do you mean here 2 dimensions?

`Just by visualizing these two matrices...` -> By visualizing the top 2 principal components...

```{r}
se@meta.data$nCount_RNA <- as.numeric(se@meta.data$nCount_RNA)
    
# FeaturePlot allows us to plot any numerical value on any latent space representation of the data. Here we can use it to color PCA by number of RNA counts per cell
nCount_RNA_pca <- FeaturePlot(
        se,
        reduction = "pca",
        features = 'nCount_RNA'
        )

nCount_RNA_pca

# DimPlot allows us to color the latent space based on categorical values per cell, like cell type or which sample the cell is from
celltype_pca <- DimPlot(
        se,
        reduction = "pca",
        group.by = 'Celltype'
        ) 

celltype_pca

# We can also plot further PCs by specifying the dimensions to plot
DimPlot(
    se,
    reduction = 'pca',
    group.by = 'Celltype',
    dims = c(3,4)
    )

donor_pca <- DimPlot(
        se,
        reduction = "pca",
        group.by = 'donor_id'
        ) + scale_color_manual(values = donor_pal)

# donor_pca
```
Based on our DimPlot of the first 2 PCs, we can see the major axes of variation in the data are 
- PC1: myeloid (negative) vs. lymphoid (positive)

> - PC2: erythroid (positive) vs. immune 

Reviewer: Add space between lines, in the notebook they are not separate bullet points
```suggestion
Based on our DimPlot of the first 2 PCs, we can see the major axes of variation in the data are 

- PC1: myeloid (negative) vs. lymphoid (positive)

- PC2: erythroid (positive) vs. immune 
```

If you asked for 100 PCs, you would see the same again for PCs 1 and 2. 

The maximum number of PCs is the total number of samples -1. so in this case, it would be nCells - 1. As we increase the number of PCs, we increase the total variance explained by the PCA in the output matrices. 

```{r, fig.width = 12, fig.height = 32}
se <- se %>% 
    RunPCA(
        npcs = 50,
        verbose = FALSE
    )

VizDimLoadings(
    se, 
    dims = c(1:2, 20:21, 40:41),
    reduction = "pca",
    # By setting balanced = TRUE, we see top loadings in each direction of each PC
    balanced = TRUE)
```

But what if we think we are missing out on some of the data by using too few PCs? Seurat has two methods for deciding the number of PCs.

### Deciding the number of Principle Components

The most popular method, the Elbow plot method, is the qualitative choice of nPCs based on deminishing returns on percent variance explained by increasing nPCs. 

> By default, the ElbowPlot shows the standard deviation of each PC. The standard deviation is the square-root of the variance. The total variance in the dataset is stored in ```se@reductions$pca@misc$total.variance``` - to obtain the percent variance explained by each PC, we can take the square of the stdev and divide it by the total variance

Reviewer: ```suggestion
By default, the `ElbowPlot` shows the standard deviation of each PC. The standard deviation is the square-root of the variance. The total variance in the dataset is stored in ```se@reductions$pca@misc$total.variance``` - to obtain the percent variance explained by each PC, we can take the square of the stdev and divide it by the total variance
```

```{r}
ElbowPlot(se, ndims = 50)

varExplained <- data.frame(varExplained = se@reductions$pca@stdev^(2)/se@reductions$pca@misc$total.variance,

> PC = seq(1:length(se@reductions$pca@stdev)))

Reviewer: ```suggestion
varExplained <- data.frame(
    varExplained = se@reductions$pca@stdev^(2) / se@reductions$pca@misc$total.variance,
    PC = seq(1:length(se@reductions$pca@stdev))
    )
```

Reviewer: just gave it some space to breathe to make it more readable

> ggplot(
>         varExplained, 
>         aes(x = PC, 
>             y = varExplained)) + 
>     geom_point() +
>     theme_classic() +
>     ylab('percent variance explained per PC')

Reviewer: ```suggestion
ggplot(varExplained,
            aes(x = PC, 
                   y = varExplained)) + 
    geom_point() +
    theme_classic() +
    ylab('percent variance explained per PC')
```

```
For the purposes of the elbow plot, we can see that the stdev is roughly equivalent to the variance explained per PC.

The Elbow plot is the simplest way to view deminishing returns on variance explained per PC. 

Then we simply cut the nPCs off at that point to prevent our PCA from being overly complex, or overfitting the data. One way we can quickly show what noise looks like in our scRNA data is by injecting noise into our dataset.

> In normal data science, typically we include PCs until we have reached some threshold of variance explained, often 85%. But if we calculate the total variance explained by each PC up to the cutoff we chose by ElbowPlot, we can see it's only 20%. This is because scRNAseq data are highly noisy with a great deal of unique variance per cell. We would have to include 100s of PCs to reach 85% variance explained.

Reviewer: maybe change: In normal data science to  In standard data science workflows...

And, keep in mind this total variance is calculated based only on the highly variable genes we chose.
```{r}

> varExplained <- data.frame(varExplained = se@reductions$pca@stdev^(2)/se@reductions$pca@misc$total.variance,
> PC = seq(1:length(se@reductions$pca@stdev)), data = 'subsampled real data')

Reviewer: ```suggestion
varExplained <- data.frame(
    varExplained = se@reductions$pca@stdev^(2) / se@reductions$pca@misc$total.variance,
    PC = seq(1:length(se@reductions$pca@stdev)),
    data = 'subsampled real data')
```

> varExplained$cumulativeSum = cumsum(varExplained$varExplained)

Reviewer: ```suggestion
varExplained$cumulativeSum <- cumsum(varExplained$varExplained)
```

ggplot(varExplained, 
       aes(x = PC, y = cumulativeSum, color = data)) +
    geom_point() +
    scale_y_continuous(limits = c(0,1)) +
    theme_classic() +
    ylab('cumulative percent variance explained per PC') +
    ggtitle('How much of the total variance in the dataset\ndid we explain with the PCs chosen by our ElbowPlot()?')

```

### Practicing PCA on a subset of the data
For those curious about just how many PC's we would need to reach that threshold of 85% variance explained, we need to subset our dataset to make visualization and comparisons reasonable.

```{r}
# select a random sample of 1000 cells
seSub <- subset(se,
                downsample = 1000) %>%
        NormalizeData() %>%
        FindVariableFeatures() %>%
        ScaleData() %>%
        RunPCA(npcs = 999)

> varExplainedsub <- data.frame(varExplained = seSub@reductions$pca@stdev^(2)/seSub@reductions$pca@misc$total.variance,
> PC = seq(1:length(seSub@reductions$pca@stdev)), data = 'subsampled real data')

Reviewer: ```suggestion
varExplainedsub <- data.frame(
    varExplained = seSub@reductions$pca@stdev^(2) / seSub@reductions$pca@misc$total.variance,
    PC = seq(1:length(seSub@reductions$pca@stdev)),
    data = 'subsampled real data')
```

ggplot(plot_data, 
       aes(x = PC, y = cumulativeSum, color = data)) +
    geom_point() +
    theme_classic() +
    facet_zoom(xlim = c(1, 20)) +
    geom_hline(yintercept = 0.85, color = 'red') +
    ylab('cumulative percent variance explained per PC') +
    ggtitle('How much of the total variance in the dataset did we explain with the PCs chosen by our ElbowPlot()?')

plot_data <- plot_data %>%
#   arrange(data, PC) %>%  data should already be sorted by PC and whether it's noise injected or not but in case you changed that <-
  group_by(data) %>%
  mutate(dropoff = c(NA, diff(varExplained)))  # Calculate differences, NA for the first entry

ggplot(plot_data, 
       aes(x = PC, y = -dropoff, color = data)) +  # Using negative dropoff to show decrease
    geom_point() +
    geom_line() +  # Optional, adds lines connecting the points
    theme_classic() +
    facet_zoom(xlim = c(1, 20)) +
    geom_hline(yintercept = 0, linetype = "dashed", color = 'blue') +  # Reference line at zero dropoff
    ylab('Dropoff in variance explained from previous PC') +
    ggtitle('Dropoff in Variance Explained for Each PC', subtitle = 'Noise-injected data vs. real scRNAseq data')

> ```

Reviewer: This chunk is hidden from the notebook, should we move it to the end just before SessionInfo?

### kNN + sNN graphs

Because these data objects can be large, the nearest neighbors graphs are normally stored in the "graphs" slot of the seurat object in a highly compact form that is not readable. We can optionally add these graphs as sets of matrices using the "return.neighbor" parameter:

```{r}
se <- FindNeighbors(
            se, 
            # We must set return.neighbor = TRUE to be able to access these matrices.
            # I wasn't able to get the shared nearest neighbors graph to save as matrices, but the kNN appears.
            return.neighbor = TRUE,
            k.param = 30
)
# Now we can print which cells are closest to one another as a matrix

> se@neighbors$RNA.nn@nn.idx[1:10,1:10]

Reviewer: ```suggestion
se@neighbors$RNA.nn@nn.idx[1:10, 1:10]
```

```
The nearest neighbors graphs are stored in the neighbors slot of the seurat object, and are stored as two matrices:
    - ```se@neighbors$RNA.nn@nn.dist```; the matrix of distances between each cell (rows) and its closest k.param neighbors (columns)
    - ```se@neighbors$RNA.nn@nn.idx```; the matrix of which cell those closest k.param cells are.
    But because these are matrices, it also stores a list of the cell names for index <-> cellname translation
    - ```se@neighbors$RNA.nn@cell.names```

```{r}

se <- FindNeighbors(
            se, 
            k.param = 30
)
```

You can imagine that if we try to display 30 lines from each cell connecting to other cells on a dataset of 60,000 cells, we would be trying to view >1,000,000 lines! This not only isn't computationally feasible for many computers, but it's also just not readable to the eye.

For educational purposes let's visualize this graph on a subset of 1000 cells:
```{r}

> seSub <- subset(se,
>                 downsample = 1000) %>%

Reviewer: ```suggestion
seSub <- subset(se, downsample = 1000) %>%
```

        NormalizeData() %>%
        FindVariableFeatures() %>%
        ScaleData() %>%
        RunPCA(npcs = 30)

> seSub <- FindNeighbors(seSub,
>                 k.param = 30,
>                 return.neighbor = TRUE
>               ) 

Reviewer: ```suggestion
seSub <- FindNeighbors(
    seSub,
    k.param = 30,
    return.neighbor = TRUE
    )
```

```

> ```{r, include = FALSE, echo = FALSE}

Reviewer: ```suggestion
Here we write a function that takes in a `Seurat` object with a KNN-graph in ```se@neighbors$RNA.nn@nn.idx``` and returns a plot with the edges between cells.
```{r, include = FALSE, echo = TRUE}
```

process_and_graph_connectivity <- function(sr) {
    all_cell_names <- colnames(sr)

    neighbors <- sr@neighbors$RNA.nn@nn.idx
    neighbors_cell_names <- apply(neighbors, 2, function(idx) all_cell_names[idx])

    edges <- data.frame()
    for (i in 1:nrow(neighbors_cell_names)) {

>     edges <- rbind(edges, data.frame(source = all_cell_names[i], target = neighbors_cell_names[i, ]))

Reviewer: ```suggestion
    edges <- rbind(
        edges, 
        data.frame(source = all_cell_names[i], target = neighbors_cell_names[i, ]))
```

    }

    edges <- subset(edges, source != target)

    pca_coords <- sr@reductions$pca@cell.embeddings[, 1:2]

    pca_coords <- as.data.frame(pca_coords)
    pca_coords$name <- rownames(pca_coords)

    graph <- tbl_graph(edges = edges, nodes = pca_coords, directed = FALSE) %>%
    activate(nodes)

    g <- ggraph(graph, layout = 'manual', x = PC_1, y = PC_2) +
        geom_edge_link(aes(alpha = 0.5), show.legend = FALSE) +
        geom_point(aes(x = PC_1, y = PC_2), color = "blue", size = 3) +
        theme_minimal() +

>         labs(title = "Cell Connectivity in PCA Space", x = "PC_1", y = "P_C2")

Reviewer: ```suggestion
        labs(title = "Cell Connectivity in PCA Space", x = "PC_1", y = "PC_2")
```

    return(g)
}
```

We've included a function here to graph the kNN graph
```{r}

> process_and_graph_connectivity(seSub)

Reviewer: ```suggestion
process_and_graph_connectivity(seSub) +
    labs(title = "K=30 Cell Connectivity in PCA Space", x = "PC_1", y = "PC_2")
```

```

We can see here that most cells are connected to those in their immediate vicinity, but cells that are more extreme along the x and y axes are also connected to more distant cells. If we decrease our k.param, this will be less pronounced.

These nearest neighbors are based on the whole PCA latent space - the distances are decided based on the "euclidean distance" between cells in ALL of the dimensions of the PCA. 

```{r}
seSub <- FindNeighbors(seSub,
                k.param = 5,
                return.neighbor = TRUE
              ) 

> process_and_graph_connectivity(seSub)

Reviewer: ```suggestion
process_and_graph_connectivity(seSub) + 
    labs(title = "K=5 Cell Connectivity in PCA Space", x = "PC_1", y = "PC_2")
```

```

> Clusters are then calculated based on the shared nearest-neighbor graph. The shared-nearest neighbor graph is a graph where the distances are replaced with the number of shared nearest neighbors between each cell and its nearest neighbors. So the  ```se@neighbors$RNA.snn@nn.idx``` matrix is the same, showing which cells are the nearest neighbors, but the ```se@neighbors$RNA.snn@nn.dist`` would be an integer number, the number of neighbors that are the same between each pair of cells.

Reviewer: ```suggestion
Clusters are then calculated based on the shared nearest-neighbor (SNN) graph. The shared-nearest neighbor graph is a graph where the distances are replaced with the number of shared nearest neighbors between each cell and its nearest neighbors. So the  ```se@neighbors$RNA.snn@nn.idx``` matrix is the same, showing which cells are the nearest neighbors, but the ```se@neighbors$RNA.snn@nn.dist`` would be an integer number, the number of neighbors that are the same between each pair of cells.  The SNN graph is used to improve the clustering robustness to outliers.
```

Let's see what happens to clusters when we modulate our k.param.
```{r, fig.width = 10, fig.height = 5}
seSub <- FindNeighbors(seSub,
                k.param = 20
              ) 

seSub <- FindClusters(seSub)
knn20 <- DimPlot(seSub, group.by= 'seurat_clusters', reduction='pca') + ggtitle('k.param = 20')

seSub <- FindNeighbors(seSub,
                k.param = 5
              ) 

seSub <- FindClusters(seSub)

knn5 <- DimPlot(seSub, group.by= 'seurat_clusters', reduction='pca') + ggtitle('k.param = 5')

(knn5 + knn20)
```

As we can see, with a higher k.param, there are fewer clusters. This is because the neighborhoods are smaller when we ask for fewer neighbors! However, in random sub-samples of the data, these clusters will be less stable because the neighbors depend on who is kept in. With a smaller k, we might get spurious clusters. Therefore, choosing k is a balance of stability and resolution on rare cell types.

```{r, fig.width = 14, fig.height = 12}

> (knn5 + knn20) / (celltype_pca + donor_pca)

Reviewer: Show cell type and donor PCA plots with the seSub object
```suggestion
celltype_pca <- DimPlot(
        seSub,
        reduction = "pca",
        group.by = 'Celltype'
        ) 
        
donor_pca <- DimPlot(
        seSub,
        reduction = "pca",
        group.by = 'donor_id'
        ) + scale_color_manual(values = donor_pal)
  
(knn5 + knn20) / (celltype_pca + donor_pca)
```

```

Even with a smaller k, we are finding multiple clusters within the authors' overarching "classical monocyte" label. However, we see a reasonable distribution of donors across them.

> If we create a stacked barplot, we can see the differences between our clusters and the author labels

Reviewer: ```suggestion
If we create Alluvial plot, we can see the differences between our clusters and the author labels
```

        geom_bar(position='stack', stat = 'identity') + 
        scale_fill_manual(values = donor_pal)
```

So we have a good distribution of samples across our normal celltypes, and we probably got all our RBCs from a single sick patient's bloody nose. In cluster 6, we have cells mostly from sick patients - these are probably some kind of activated cells.

We can get the best idea of what's going on there with a differential expression, but let's save that for later.

## Integration with Harmony

Switching gears a little bit before we get ahead of ourselves making clusters, let's double check we don't have any batch effects at play that may cause batch-specific clustering. 

Usually our batch effects will show up as some vector of variation across a metadata variable. This means they will usually show up in our PCA as a PC, so we can check for them by correlating our PCs with our metadata!
```{r, fig.width = 14, fig.height = 14}

plot_batch_effect_heatmap <- function(seurat_object, 
                                     latent_space = 'pca', 
                                     num_pcs = 20, 
                                     metadata_columns_to_include = NULL,

>                                      plot_title = "Correlation Heatmap of Metadata and PCA Components") {

Reviewer: ```suggestion
plot_batch_effect_heatmap <- function(
    seurat_object,
    latent_space = 'pca', 
    num_pcs = 20, 
    metadata_columns_to_include = NULL,
    plot_title = "Correlation Heatmap of Metadata and PCA Components") {
```

    metadata <- seurat_object@meta.data
    metadata <- metadata[, metadata_columns_to_include, drop = FALSE]
    pca_data <- Embeddings(seurat_object, latent_space)[, 1:num_pcs] # pull the PCs we want to check

    categorical_cols <- sapply(metadata, function(col) is.factor(col) || is.character(col))
    

>     metadata[categorical_cols] <- lapply(metadata[categorical_cols], as.factor) # convert character columns to factors for one-hot encoding

Reviewer: ```suggestion
    # convert character columns to factors for one-hot encoding
    metadata[categorical_cols] <- lapply(metadata[categorical_cols], as.factor)
```

    #1hot without removing redundant dummies

>     metadata_encoded <- fastDummies::dummy_cols(metadata, remove_first_dummy = FALSE, remove_selected_columns = TRUE)

Reviewer: ```suggestion
    metadata_encoded <- fastDummies::dummy_cols(
    metadata,
    remove_first_dummy = FALSE,
    remove_selected_columns = TRUE)
```

    # calculate the correlation matrix
    correlation_matrix <- cor(combined_data, use = "pairwise.complete.obs")

    # Subset the correlation matrix to include only metadata and PCA components
    metadata_columns <- colnames(metadata_encoded)
    pca_columns <- colnames(pca_data)
    subset_correlation_matrix <- correlation_matrix[metadata_columns, pca_columns, drop = FALSE]
    
    # Check for any NA values in the subset correlation matrix and replace them with 0
    subset_correlation_matrix[is.na(subset_correlation_matrix)] <- 0
    
    # Create the heatmap using corrplot
    corrplot(subset_correlation_matrix, method = "circle", tl.col = "black", tl.srt = 45,
             number.cex = 0.7, tl.cex = 0.7,
             main = plot_title)
}

plot_batch_effect_heatmap(se, 
                        metadata_columns_to_include = c('Age', 
                                                        'sex', 
                                                        "donor_id", 
                                                        "Number of UMI", 
                                                        "disease",
                                                        "Severity"))

> 

Reviewer: ```suggestion
plot_batch_effect_heatmap(
    se, 
    metadata_columns_to_include = c('Age', 'sex',  "donor_id", "Number of UMI", 
                                    "disease", "Severity"))
```

pcs1_2 <- DimPlot(se,
        reduction = 'pca',
        group.by = 'donor_id') + 
        scale_color_manual(values = donor_pal) +
        NoLegend()

pcs1_8 <- DimPlot(se,
        reduction = 'pca',
        group.by = 'donor_id',
         dims = c(1,8)) + 
        scale_color_manual(values = donor_pal) 

(pcs1_2 + pcs1_8) / (h1_2 + h1_8)
```

By comparing the PCs directly, we can already see the Flu 1 effect has disappeared from PC_8. 
Typically we would next want to check the effects of this change on the correlation structure of our whole dataset.
We can do this by plotting again the correlation of metadata covariates

```{r, fig.width = 14, fig.height = 14}
plot_batch_effect_heatmap(se, 
                        latent_space = 'harmony',
                        metadata_columns_to_include = c('Age', 
                                                        'sex', 
                                                        "donor_id", 
                                                        "Number of UMI", 
                                                        "disease",
                                                        "Severity"),

>                         plot_title = "Correlation Heatmap of Metadata and Harmony latent variables")

Reviewer: ```suggestion
plot_batch_effect_heatmap(
    se, 
    latent_space = 'harmony',
    metadata_columns_to_include = c('Age', 'sex', "donor_id", "Number of UMI",
                                    "disease", "Severity"),
    plot_title = "Correlation Heatmap of Metadata and Harmony latent variables")
```

> ```

Reviewer: ```suggestion

```

## Session Info
```{r}
sessionInfo()
```

```
